README.md
deepmac
=======

The DeepMac project is an attempt to archive indefinitely all OUI records and provide additional metadata related to OUI registrations. Data can be used by IT professionals for asset management and by security professionals for recon and assessment.

The code here is a long-term work-in-progress, with a focus on data and results over code quality, maintainability or efficiency.
Original project started in 2008 and has moved slowly. A full redesign is in progress that focuses on a better way to record metadata about OUI registrations. Future features will allow for search engines that can find new metadata to add to the journals. See below for code overview.

IMPORTANT NOTES
---------------
Due to the nature of this project, errors in the original code led to irregularities in the database, such as duplicate Company entries, incorrect country codes and other artifacts. The author is aware of these, and intends to retire all the old code and data when the redesign is complete and functional.

While the data for DeepMac is shared under an open license, the original base files in their curated form are considered to be under an archival copyright and are not currently available for distribution. They may be available for purchase/licensing at a future date, this has yet to be determined. You may import the SQL provided into your own instance and then start running your own archiving process of course.


Requirements (Original codebase)
--------------------------------
* A webserver with PHP and MySQL support
* PHP 5.x (4.x may actually work fine)
* Perl 5.x
* MySQL 14 (older versions will almost definitely work)
* Perl modules
 * DBD::mysql
 * Data::Dumper (opt)
 * Geo::PostalAddress (opt)
* PHP modules
 * DB                                        1.7.13  stable
 * DB_DataObject                             1.9.3   stable
 * DB_Table                                  1.5.6   stable
 * Date                                      1.4.7   stable
 * HTML_Common                               1.2.5   stable
 * HTML_QuickForm                            3.2.11  stable
 * HTML_Table                                1.8.3   stable
 * MDB2                                      2.4.1   stable
 * MDB2_Driver_mysql                         1.4.1   stable
 * PHPUnit                                   1.3.2   stable
 * Pager                                     2.4.8   stable
 * Structures_DataGrid                       0.9.0   beta
 * Structures_DataGrid_DataSource_Array      0.1.4   beta
 * Structures_DataGrid_DataSource_DB         0.1.1   beta
 * Structures_DataGrid_DataSource_DBQuery    0.1.11  beta
 * Structures_DataGrid_DataSource_DBTable    0.1.7   beta
 * Structures_DataGrid_DataSource_DataObject 0.2.1   beta
 * Structures_DataGrid_DataSource_MDB2       0.1.11  beta
 * Structures_DataGrid_Renderer_Console      0.1.1   beta
 * Structures_DataGrid_Renderer_HTMLSortForm 0.1.3   beta
 * Structures_DataGrid_Renderer_HTMLTable    0.1.5   beta
 * Structures_DataGrid_Renderer_Pager        0.1.3   beta
 * Structures_Graph                          1.0.3   stable
 * XML_Util                                  1.2.1   stable

Try and get all requirements together and functioning. Create a MySQL database to hold the original project data. You can initialize it using the dictionary files as a guide, or do a full import of the most recently available SQL dump.
Once a DB has been created and initialized, set-up credentials for it. A set of creds for read/write access to be used by the perl scripts for data loading or manual manipulation, and a read-only set to be used by the PHP search interface.
Add a file named "creds.txt" with the username and password to use for updating the DB. Format is a single line with "username	password" (tab whitespace).
Run and test the various perl scripts to make sure they execute and function correctly. You will need to fix hard-coded paths, put in the correct hostname for the DB backend, etc.
Establish cron jobs to run daily to perform the necessary archiving and updated tasks. Here's an example crontab:

	# Archive IEEE files
	00 01 * * *     ~USERDIR/deepmac/ouiarchive.pl >> ~USERDIR/archive.log
	15 01 * * *     ~USERDIR/deepmac/reboot/ouiarchive.pl >> ~USERDIR/deepmac/reboot/archive.log

	# Generate a master CSV files of dates associated with OUIs based on the archive
	00 02 * * *     ~USERDIR/deepmac/gen-ouidates.pl > ~USERDIR/deepmac/kb/oui-dates.csv

	# Run a database load, using the archive of IEEE OUI files
	00 03 * * *     ~USERDIR/deepmac/dbload.pl >> ~USERDIR/dbload.log

	# Generate some stats from the database for public consumption
	00 04 * * *     ~USERDIR/deepmac/stats.pl > ~USERDIR/deepmac/stats.txt

For the search interface, the PHP code will need to be executable by your webserver, so make sure you put together an appropriate and secure configuration. Deepmac has always used Apache on Linux but on a shared hosting provider.
Test the search interface to make sure it can access the DB and produce results.



Original Code
-------------
	.
	|-- bits		<-- Sub-dir for some PHP code
	|-- dbload.pl		<-- Perl script to load MySQL database from a master OUI data file
	|-- deepmacdict.html	\__ Database dictionary for MySQL back-end
	|-- deepmacdict_files	/
	|-- details.php		<-- PHP code to display detailed info on an OUi from backend
	|-- dump.sh		<-- Simple Shell script to generate MySQL dumps (for cron)
	|-- gen-ouidates.pl	<-- Perl script that generates a master OUI list with dates, from an archive of OUI files
	|-- index.php		<-- Main PHP index page for Deepmac search
	|-- oui2csv.pl		<-- Perl script to convert an IEEE-standard OUI text file to a tab-delimited format
	|-- ouiarchive.pl	<-- Perl script to check for and archive IEEE OUI files (run as daily cron job)
	|-- search.php		<-- PHP search code for DeepMac search interface
	|-- sqldump.sql		<-- A MySQL dump of the database, includes metadata not found in the OUI files
	|-- stats.pl		<-- Perl script to give some stats on the database backend
	`-- workshop		<-- Sub-dir with one-offs
	    |-- 20090427	<-- Restoration of a specific archived OUI dataset
	    |-- bacnet-map.pl	<-- Non-working attempt to map bacnet IDs to OUI registrations
	    |-- nmaparchive	<-- Archive of NMap SVN codes, for finding old OUI registry datasets
	    |-- php		<-- PHP code archive
	    `-- unbolted.net	<-- Archive of unbolted.net data related to Wireless profiles of OUI registrations


Project Reboot
--------------
	|-- reboot
	|   |-- deepmac_connector.py	<-- DeepMac connector class. Connect to a DeepMac journal (currently filesystem only)
	|   |-- deepmac_import.py	<-- Python script to import OUI registry data into DeepMac journal
	|   |-- deepmac_manager.py	<-- DeepMac manager class. Mange connections, journal operations, etc.
	|   |-- deepmac_record_class.py	<-- DeepMac record class. Defines journal record format as an object, manipulates record entries, etc.
	|   |-- dmimport.cfg		<-- Config file for deepmac_import.py
	|   |-- gen-ouidates.pl		<-- Perl script that generates a master OUI list with dates, from an archive of OUI files
	|   |-- journal			<-- Sub-dir for holding DeepMac journal entries (for filesystem mode)
	|   |-- kb			<-- Sub-dir for holding IEEE OUI archive
	|   |-- oui2csv.pl		<-- Perl script to convert an IEEE-standard OUI text file to a tab-delimited format
	|   |-- ouiarchive.pl		<-- Perl script to check for and archive IEEE OUI files (run as daily cron job)


CHANGELOG
2019-06-13
	- Added a changelog!
	  (Note: You can find summaries of changes in both scripts in the comments at the top)
	- Updated to use most recent fixes from May 2019 for some long sought-after bugs with importing
	  IEEE source files into the journal.
	- Some misc. bug fixes for changes in IEEE distribution of registry files (location, format, etc)
	- verify() function in deepmac_record_class.py no longer fails on blank orgnames
	- deepmac_import.py now does case-insensitive comparisons between records in the journal. This
	  eliminates logging of trivial changes from the IEEE source.



search2.php
<?php
#  - Main Index page
#
# Author: Jeff Mercer <jedi@jedimercer.com>
# Written: 07/07/09
# Updated: 02/14/13
# Purpose: This particular page is the main index where the
#          user specifies initial search parameters.

# Change include search path to include the location of locally installed PEAR components
ini_set('include_path', ini_get('include_path').PATH_SEPARATOR."/home/riffer/pear/php");

# For doing rendering results of searches
require 'Structures/DataGrid.php';

# Comment out to disable error reporting
#	error_reporting(E_ALL);
#	ini_set('display_errors', 1);
#	PEAR::setErrorHandling(PEAR_ERROR_DIE);

# Initialization and configuration
$title="DeepMac - OUI Search - Search Results";
include("bits/access.php");
include("bits/header.php");





<CENTER>
<H1>DeepMac</H1>
<H2>OUI Search Interface</H2>
</CENTER>

<FONT SIZE=+1>
This is an alpha search interface to the DeepMac database, which itself is a grandiose project in its infancy. Do not be 
surprised if something does not work, the search page disappears, the site gets bogged down or that the data returned is not 
what you expected. For more details on what DeepMac is, see the <A HREF="http://www.jedimercer.com/projects/48-deepmac.html">project page</A>. 
For more inforation on what an OUI is, see the <A HREF="http://standards.ieee.org/regauth/registries.html">IEEE website</A> 
as well as <A HREF="http://en.wikipedia.org/wiki/Organizationally_Unique_Identifier">this Wikipedia article</A>.
</FONT>
<P>

<DIV ALIGN=CENTER>
<HR WIDTH=35%>

<?php
# Attempt connection to database server
if (!($DBI=mysql_connect($DBHost, $DBUser, $DBPass))) {
  # Failed to connect, report error and quit.
  echo "<P>\n<B>Error!</B> Unable to connect to database server on $DBHost as $DBUser\n";
  include("footer.php");
  die();

  # Attempt selection of specified database instance
} elseif (!(mysql_select_db($DBName, $DBI))) {
  # Couldn't select the database, report error and quit.
  echo "<P>\n<B>Error!</B> Unable to select database $DBName on $DBHost\n";
  include("footer.php");
  die();
} else {
  # Everything checks out ok, so close the DB connection
  mysql_close($DBI);
}
?>

<EM>
Leave a field blank to match all records of that type.<BR>
All search terms are wildcarded automatically and are case-insensitive.<BR>
Use a percent sign (%) for wildcard matching inside a search term.
</EM>
<HR WIDTH=35%>

<TABLE FRAME=BORDER TITLE="DeepMac Search Form">
<CAPTION>DeepMac Search Form</CAPTION>

<FORM ACTION="search.php" METHOD=GET>

<TR>
<TD>Date:</TD>
<TD><INPUT TYPE="text" MAXLENGTH=10 SIZE=12 NAME="date">
<SUB><I>(YYYY-MM-DD, i.e. 2002-03-22)</I></SUB></TD>
</TR>

<TR>
<TD>MAC Address:</TD>
<TD><INPUT TYPE="text" MAXLENGTH=17 SIZE=19 NAME="macadd">
<SUB><I>(Seperators such as colons, hyphens or periods are optional)</I></SUB>
</TD>
</TR>

<TR>
<TD>Company Name:</TD>
<TD><INPUT TYPE="text" MAXLENGTH=128 SIZE=32 NAME="comp"></TD>
</TR>

<TR>
<TD>Device Type:</TD>
<TD><INPUT TYPE="text" MAXLENGTH=128 SIZE=32 NAME="dev"></TD>
</TR>

<TR>
<TD>Results per page:</TD>
<TD><INPUT TYPE="text" MAXLENGTH=4 SIZE=4 NAME="numresults" VALUE=50></TD>
</TR>

</TABLE>

<P>

<INPUT TYPE="reset" VALUE="Clear Form">
&nbsp;
<INPUT TYPE="submit" VALUE="Search">
</FORM>

<?php
include("bits/footer.php");
?>

</DIV>
</BODY>
</HTML>
































# Simple function for callback from Structures_DataGrid_Column
function DMCompany($row) {
  extract($row);

  return "{$record['compname']}<BR>{$record['country']}";
}

# Simple function for callback from Structures_DataGrid_Column
function DMOUI($row) {
  extract($row);

  return "<A HREF=\"details.php?macadd={$record['prefix']}&numresults=10\">{$record['prefix']}</A>";
}

# If this is a submitted form page, we do a search
# Otherwise, display entry form and exit.
if (isset($_POST['submit'] == "Submit") {
} else {
}


# Get our passed-in parameters and sterilize against attempts to do SQL
# injection or other nefarious deeds.
$date		= escapeshellcmd(trim($_GET['date']));
$macadd		= escapeshellcmd(trim($_GET['macadd']));
$comp		= escapeshellcmd(trim($_GET['comp']));
$dev		= escapeshellcmd(trim($_GET['dev']));
$numresults	= escapeshellcmd(trim($_GET['numresults']));

#
# Do defaults for unspecified parameters
#

# Pagination setting defaults to 50 if it wasn't specified
if ($numresults == "" || empty($numresults)) { $numresults=50; }

#
# Perform data validation checks, abort if any of them fail.
#

# Check host length
if (strlen($comp) > 128) {
  echo "Company search parameter $comp is too long.\n<P>\n";
  echo "<CENTER><A HREF=\"index.php\">New Search</A>";
  include("bits/footer.php");
  die();
}

# Check MAC address string length
if (strlen($macadd) > 17) {
  echo "OUI search parameter $macadd is too long.\n<P>\n";
  echo "<CENTER><A HREF=\"index.php\">New Search</A>";
  include("bits/footer.php");
  die();
}

# Check date string length
if (strlen($date) > 10) {
  echo "Date search parameter $date is too long.\n<P>\n";
  echo "<CENTER><A HREF=\"index.php\">New Search</A>";
  include("bits/footer.php");
  die();
}

# Make sure $numesults variable is numeric and a valid length
if (is_numeric($numresults) != 1 or strlen($numresults) > 4) {
  echo "Number of results per page value $numresults is invalid.\n<P>\n";
  echo "<CENTER><A HREF=\"index.php\">New Search</A>";
  include("bits/footer.php");
  die();
}

# Instantiate the DataGrid object
$DGrid=& new Structures_DataGrid($numresults);

# Define the database connection string
$DBConn=array("dsn" => "mysql://$DBUser:$DBPass@$DBHost/$DBName");

$DGrid->setDefaultSort(array('date'=>'ASC'));

# Remove seperators from MAC field, then truncate anything after the OUI
$oui=substr(preg_replace("/[^A-Za-z0-9%]*/", "", $macadd), 0, 6);

# Put any search parameters in the stack
if ($oui != "")  { $Where[]="tb_OUI.prefix LIKE '%$oui%'"; }
if ($date != "") { $Where[]="tb_OUI.date LIKE '%$date%'"; }
if ($comp != "") { $Where[]="tb_company.compname LIKE '%$comp%'"; }
if ($dev != "")  { $Where[]="tb_device.devname LIKE '%$dev%'"; }

##############################################################################

# Prep actual search query.
$SQLState='
SELECT tb_OUI.prefix, tb_OUI.date, tb_company.compname, tb_company.country, 
tb_media.medianame, tb_device.devname, tb_model.modelname
  FROM tb_OUI
  LEFT JOIN tb_company
  ON tb_OUI.comp_id = tb_company.comp_id
  LEFT JOIN tb_media
  ON tb_OUI.media_id = tb_media.media_id
  LEFT JOIN tb_device
  ON tb_OUI.dev_id = tb_device.dev_id
  LEFT JOIN tb_model
  ON tb_OUI.model_id = tb_model.model_id
';

# Add WHERE clause if there's one or more search parameters
if (count($Where)) {
  $WC=join(" AND ", $Where);
  $SQLState=$SQLState." WHERE ($WC)";
}

# Execute the search query
$sql_result=$DGrid->bind($SQLState, $DBConn);

# Print any binding/querying errors, if any
if (PEAR::isError($sql_result)) {
  echo "ERROR! Could not bind to database, or query failed! Original error message, if any, follows: ###> ";
  echo $sql_result->getMessage();
  echo " <### Giving up.";
  die;
}


# If there were no results, give an error
if ($DGrid->getRecordCount()==0) {
#  print "oui = $oui<br>\n";
#  print "date = $date<br>\n";
#  print "comp = $comp<br>\n";
#  print "dev = $dev<br>\n";

?>
<CENTER>No results were found for this search.<P>
<A HREF="index.php">New Search</A>
<?php
  include("bits/footer.php");
  die();
# Otherwise...
} else {
  # Start displaying the results
  $row_count=0;
?>

<CENTER>
<H1>Matching DeepMac Entries</H1>
<P>
<A HREF="index.php">New Search</A>
<P>

<?php
# Create our columns without speciality formatting
$DGrid->generateColumns(array(
#  'prefix' => 'OUI',
  'date'   => 'Date',
));

# Manually add columns with special formatting
# And yes, this requires you specify a function name to do the formatting.
# And no, you can't just pass the field name into that function. Because this is the buttfuck way to do it
$DGrid->addColumn(new Structures_DataGrid_Column("OUI", "prefix", "prefix", null, null, "DMOUI()"));
$DGrid->addColumn(new Structures_DataGrid_Column("Company", "compname", "compname", null, null, "DMCompany()"));
$DGrid->addColumn(new Structures_DataGrid_Column("Media Type", "medianame", "medianame", array('align' => 'center'), null));
$DGrid->addColumn(new Structures_DataGrid_Column("Device Class", "devname", "devname", array('align' => 'center'), null));
$DGrid->addColumn(new Structures_DataGrid_Column("Device Model", "modelname", "modelname", array('align' => 'center'), null));

# Specify rendering options
$DGrid->setRendererOptions(array(
  'sortIconASC'         => '&uArr;',
  'sortIconDESC'        => '&dArr;',
  'headerAttributes'    => array('bgcolor' => '#E3E3E3'),
  'evenRowAttributes'   => array('bgcolor' => '#C6C6C6'),
));

# Specify border around table
$Rend=$DGrid->getRenderer();
$Rend->setTableAttribute('border', 1);

# Print the DataGrid with the default renderer (HTMLTable)
$Result=$DGrid->render(DATAGRID_RENDER_PAGER);
$Result=$DGrid->render();
$Result=$DGrid->render(DATAGRID_RENDER_PAGER);

?>

<P>
<A HREF="index.php">New Search</A>
</CENTER>

<?php
}

# End of the line
include("bits/footer.php");
?>
